### PR TITLE
CXP-943: Add catalog structure Scope Mapper

### PR DESCRIPTION
**Pull request by willy-ahva**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

**Review by masacc — APPROVED** [review]

File: src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            'pim_api_attribute_group_edit',
+            'pim_api_family_edit',
+            'pim_api_family_variant_edit',
+        ],
+    ];
+
+    private const SCOPE_MESSAGE_MAP = [
+        self::SCOPE_READ_CATALOG_STRUCTURE => [
+            'icon' => 'catalog_structure',
+            'type' => 'read',
```
Comment (masacc) [inline src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php:37]:
```suggestion
            'type' => 'view',
```

File: src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            'entities' => 'catalog_structure',
+        ],
+        self::SCOPE_WRITE_CATALOG_STRUCTURE => [
+            'icon' => 'catalog_structure',
+            'type' => 'write',
```
Comment (masacc) [inline src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php:42]:
```suggestion
            'type' => 'edit',
```

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/services.yml [hunk 3]
```diff
@@ -90,3 +90,7 @@ services:
             - '@oro_filter.filter_utility'
         tags:
             - { name: oro_filter.extension.orm_filter.filter, type: datagrid_attribute_family_filter }
+
+    Akeneo\Pim\Structure\Component\Security\CatalogStructureScopeMapper:
+        tags:
+            - { name: pim_api.security.scope_mapper, entity: 'catalog_structure' }
```
Comment (masacc) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/services.yml:96]:
```suggestion
            - { name: pim_api.security.scope_mapper }
```

File: src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        ],
+    ];
+
+    private const SCOPE_HIERARCHY = [
+        self::SCOPE_WRITE_CATALOG_STRUCTURE => [
+            self::SCOPE_READ_CATALOG_STRUCTURE,
+        ],
+    ];
+
+    public function getAllScopes(): array
```
Comment (masacc) [inline src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php:53]:
```suggestion
    public function getScopes(): array
```

File: src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Structure\Component\Security;
+
+use Akeneo\Tool\Component\Api\Security\ScopeMapperInterface;
```
Comment (masacc) [inline src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php:7]:
```suggestion
use Akeneo\Connectivity\Connection\Infrastructure\Apps\Security\ScopeMapperInterface;
```

File: src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    public function getAcls(string $scopeName): array
+    {
+        if (!\array_key_exists($scopeName, self::SCOPE_ACL_MAP)) {
+            return [];
+        }
+
+        return self::SCOPE_ACL_MAP[$scopeName];
+    }
+
+    public function getMessage(string $scope): array
```
Comment (masacc) [inline src/Akeneo/Pim/Structure/Component/Security/CatalogStructureScopeMapper.php:70]: `$scope` => `$scopeName` everywhere in this file

File: tests/back/Pim/Structure/Specification/Component/Security/CatalogStructureScopeMapperSpec.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Specification\Akeneo\Pim\Structure\Component\Security;
+
+use Akeneo\Pim\Structure\Component\Security\CatalogStructureScopeMapper;
+use Akeneo\Tool\Component\Api\Security\ScopeMapperInterface;
```
Comment (masacc) [inline tests/back/Pim/Structure/Specification/Component/Security/CatalogStructureScopeMapperSpec.php:8]:
```suggestion
use Akeneo\Connectivity\Connection\Infrastructure\Apps\Security\ScopeMapperInterface;
```

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/services.yml [hunk 8]
```diff
@@ -90,3 +90,7 @@ services:
             - '@oro_filter.filter_utility'
         tags:
             - { name: oro_filter.extension.orm_filter.filter, type: datagrid_attribute_family_filter }
+
+    Akeneo\Pim\Structure\Component\Security\CatalogStructureScopeMapper:
```
Comment (masacc) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/services.yml:94]: rather declare the service in a `scope_mappers.yml` ?

File: tests/back/Pim/Structure/Specification/Component/Security/CatalogStructureScopeMapperSpec.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+{
+    public function it_is_a_catalog_structure_scope_mapper(): void
+    {
+        $this->shouldHaveType(CatalogStructureScopeMapper::class);
+        $this->shouldImplement(ScopeMapperInterface::class);
+    }
+
+    public function it_provides_all_scopes(): void
+    {
+        $this->getAllScopes()->shouldReturn([
```
Comment (masacc) [inline tests/back/Pim/Structure/Specification/Component/Security/CatalogStructureScopeMapperSpec.php:21]:
```suggestion
        $this->getScopes()->shouldReturn([
```

File: tests/back/Pim/Structure/Specification/Component/Security/CatalogStructureScopeMapperSpec.php [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+            'pim_api_family_variant_edit',
+        ]);
+    }
+
+    public function it_does_not_provide_acl_if_an_unknown_scope_is_given(): void
+    {
+        $this->getAcls('unknown_scope')->shouldReturn([]);
+    }
+
+    public function it_provides_message_that_correspond_to_read_catalog_structure_scope(): void
```
Comment (masacc) [inline tests/back/Pim/Structure/Specification/Component/Security/CatalogStructureScopeMapperSpec.php:52]:
```suggestion
    public function it_provides_message_that_corresponds_to_the_read_catalog_structure_scope(): void
```
same for others

File: tests/back/Pim/Structure/Specification/Component/Security/CatalogStructureScopeMapperSpec.php [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        $this->getAcls('write_catalog_structure')->shouldReturn([
+            'pim_api_attribute_edit',
+            'pim_api_attribute_group_edit',
+            'pim_api_family_edit',
+            'pim_api_family_variant_edit',
+        ]);
+    }
+
+    public function it_does_not_provide_acl_if_an_unknown_scope_is_given(): void
```
Comment (masacc) [inline tests/back/Pim/Structure/Specification/Component/Security/CatalogStructureScopeMapperSpec.php:47]:
```suggestion
    public function it_throws_an_exception_if_an_unknown_scope_is_given(): void
```
same everywhere

**Review by RiffVentura — APPROVED** [review]

**Pull request merged**